### PR TITLE
LPS-151149 Implement missing features in Sample FDS module

### PR DESCRIPTION
**Pull request by beltranrengifo**
[Jira issue](https://issues.liferay.com/browse/LPS-151149)

## Motivation

Adding some basic features missing in the Sample FDS.

## How to
- Just deploy the sample module and place it in a page using the page editor. You'll find it under widgets
 
image

## What's new?
- Sorting for title column
- Tabs for showcasing multiple fds
- Multiple and Single selection, with no logic associated to the actions, yet.

### CODE REVIEW

**Comment by beltranrengifo** [comment]
ci:stop

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by beltranrengifo** [comment]
ci:test:relevant

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by beltranrengifo** [comment]
ci:test:relevant

**Comment by beltranrengifo** [comment]
ci:stop

**Comment by beltranrengifo** [comment]
ci:test:sf

**Comment by beltranrengifo** [comment]
ci:test:relevant

**Review by markocikos — COMMENTED** [review]
@beltranrengifo please see comments inline.

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/view.jsp [hunk 1]
```diff
@@ -20,21 +20,111 @@
...
+
+			clickedTabNavItemLink.classList.add('active');
+			document
+				.getElementById(
+					clickedTabNavItemLink.getAttribute('href').replace('#', '')
+				)
+				.classList.add('active', 'show');
+		});
+	});
+</script>
```
[the remark this answers is not in the document]
Comment (beltranrengifo) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/view.jsp:130]: I searched a bit for `clay:tabs` taglib, with no luck. Then I read in [Clay](https://clayui.com/docs/get-started/using-clay-in-jsps.html) it's not implemented and I didn't know about that `liferay:ui` taglibs, that's why I went for this solution. Thanks for pointing it out! 🎉 

I give this taglib a try and let you know how it feels.

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/view.jsp [hunk 2]
```diff
@@ -20,21 +20,111 @@
...
-/>
\ No newline at end of file
+<section id="<portlet:namespace />fdsSample">
+	<clay:container-fluid
+		cssClass="container-fluid container-fluid-max-xl my-4"
+	>
+		<ul class="nav nav-underline" role="tablist">
+			<li class="nav-item">
+				<a aria-controls="navUnderlineMultiple" aria-selected="true" class="active nav-link" data-toggle="tab" href="#navUnderlineMultiple" id="navUnderlineMultipleTab" role="tab">
+					Multiple selection
```
[the remark this answers is not in the document]
Comment (beltranrengifo) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/view.jsp]: Yes, I agree. Indeed I used this names but was hoping to discuss it with you.
I like the approach 1, with generic names and listing the expected features in each case.

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/view.jsp [hunk 3]
```diff
@@ -20,21 +20,111 @@
...
+					fdsActionDropdownItems="<%= fdsSampleDisplayContext.getFDSActionDropdownItems() %>"
+					formId="fm"
+					id="<%= FDSSampleFDSNames.FDS_SAMPLES %>"
+					itemsPerPage="<%= 20 %>"
+					namespace="<%= liferayPortletResponse.getNamespace() %>"
+					pageNumber="<%= 1 %>"
+					portletURL="<%= liferayPortletResponse.createRenderURL() %>"
+					propsTransformer="js/SampleFDSPropsTransformer"
+					selectedItemsKey="id"
+					selectionType="single"
```
[the remark this answers is not in the document]
Comment (beltranrengifo) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/view.jsp:94]: I think same way, I can't find the utility and I don't get why this is even implemented.
I tried to find the usages of `selectionType="single"` too with no luck.
I'll take a look again just confirm, then remove it

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/view.jsp (hunk 2 above)
[the remark this answers is not in the document]
Comment (beltranrengifo) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/view.jsp]: Uhmm maybe if we end up with many examples, giving them a more explicit name will help to identify each tab, like you said in case 2...

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/view.jsp (hunk 1 above)
[the remark this answers is not in the document]
Comment (beltranrengifo) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/view.jsp]: Pushed!
It seems to work fine and I saved lots of code 😃

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/minimal.jsp [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+		id="<%= FDSSampleFDSNames.FDS_SAMPLES %>"
+		itemsPerPage="<%= 20 %>"
+		namespace="<%= liferayPortletResponse.getNamespace() %>"
+		pageNumber="<%= 1 %>"
+		portletURL="<%= liferayPortletResponse.createRenderURL() %>"
+		propsTransformer="js/SampleFDSPropsTransformer"
+		showManagementBar="<%= false %>"
+		showPagination="<%= false %>"
+		style="fluid"
+	/>
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/minimal.jsp:42]: If we have a showcase of a minimal configuration, we should reduce amount of attributes to minimum. This is how I think API should be, it may not match implementation:
- `additionalProps` - not required, no additional props by default
- `apiURL` - required
- `customViewsEnabled` - not required, disabled by default
- `fdsActionDropdownItems` - not required, no actions by default
- `formId` - not required, `fm` by default
- `id` - required
- `itemsPerPage` - not required, 20 by default
- `namespace` - not required, portlet namespace by default
- `pageNumber` - not required, 1 by default
- `portletURL` - not required. Do we need this? I believe we can get this as an internal mechanism in the tag class.
- `propsTransformer` - not required, no props transformations by default
- `showManagementBar` - not required, true by default
- `showPagination` - not required, true by default
- `style` - not required, `style` by default. Sidenote - do we need this? As far as I know, we plan on having just one set of styles.

To summarize, a showcase of minimal configuration should have only `apiURL` and `id`.

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/multiple_selection.jsp [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+		id="<%= FDSSampleFDSNames.FDS_SAMPLES %>"
+		itemsPerPage="<%= 20 %>"
+		namespace="<%= liferayPortletResponse.getNamespace() %>"
+		pageNumber="<%= 1 %>"
+		portletURL="<%= liferayPortletResponse.createRenderURL() %>"
+		propsTransformer="js/SampleFDSPropsTransformer"
+		selectedItemsKey="id"
+		selectionType="multiple"
+		style="fluid"
+	/>
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/multiple_selection.jsp:43]: If we have a tab that showcases only multiple selection, we should remove all the noise attributes. It should be minimal configuration + anything required for multiple selection. I believe that's only `selectionType`.

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/basic.jsp [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+--%>
+
+<%@ include file="/init.jsp" %>
+
+<form action="<%= fdsSampleDisplayContext.getPortletURL() %>" method="post" name="fm">
+	<input name="<%= Constants.CMD %>" type="hidden" value="<%= Constants.DELETE %>" />
+	<input name="redirect" type="hidden" value="<%= String.valueOf(fdsSampleDisplayContext.getPortletURL()) %>" />
```
Comment (markocikos) [inline modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/basic.jsp:21]: This form around the FDS does not look right. Shouldn't all actions be `fetch` requests? :thinking:  In theory, we could make the form work, but wouldn't ajax be a better tool for the job?

**Review by markocikos — COMMENTED** [review]
@beltranrengifo Left a few comments inline, with clarification what I meant by `Minimal Configuration`.

File: modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/minimal.jsp (hunk 4 above)
Comment (beltranrengifo) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/minimal.jsp:42]: Yes, you're right.
This are the mandatory attributes:
```
<frontend-data-set:headless-display
		apiURL="<%= fdsSampleDisplayContext.getAPIURL() %>"
		id="<%= FDSSampleFDSNames.FDS_SAMPLES %>"
		itemsPerPage="<%= 20 %>"
		namespace="<%= liferayPortletResponse.getNamespace() %>"
		pageNumber="<%= 1 %>"
		portletURL="<%= liferayPortletResponse.createRenderURL() %>"
/>
```

Removing `additionalProps` breaks the component. If I understand correctly the component, we should have different `SampleTableFDSView.java` views, each of one with it's proper schema, right?

Comment (beltranrengifo) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/minimal.jsp:42]: I agree with you that we should make it easier, the less mandatory attributes the better. I wonder where is the right place to pass the default values, the tag?

Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/minimal.jsp:42]: > I agree with you that we should make it easier, the less mandatory attributes the better. I wonder where is the right place to pass the default values, the tag?

Generally, it's better to have default on the frontend component prop, as we are supporting direct usage (not over JSP tag). For some attibutes, tag class default may be a better fit, if they have a backend component. For example, `namespace`.

Comment (markocikos) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/minimal.jsp:42]: > Removing `additionalProps` breaks the component. If I understand correctly the component, we should have different `SampleTableFDSView.java` views, each of one with it's proper schema, right?

I'm not sure what you mean. `additionalProps` is related to `propsTransformer`. If lack of it breaks the component, that's a bug. Are you sure that it does? If it did, almost all usages FDS in DXP would not work.

Comment (beltranrengifo) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/minimal.jsp:42]: > * `style` - not required, `style` by default. Sidenote - do we need this? As far as I know, we plan on having just one set of styles.

I think is useful the option to have the FDS boxed or fullwidth, but it's something can be easily done using Clay layout tools/components, so we could get rid of it... 
We can break some things however...

Comment (beltranrengifo) [reply modules/apps/frontend-data-set/frontend-data-set-sample-web/src/main/resources/META-INF/resources/partials/minimal.jsp:42]: > > Removing `additionalProps` breaks the component.

Sorry @markocikos I meant `propsTransformer`...
This: "Removing propsTransformer breaks the component because it expects a content renderer, uhmmm"

> Each FDS instance should have its own view definition, yes. FDS id must match frontend.data.set.name property.
Great, understood

**Comment by beltranrengifo** [comment]
@markocikos I'm closing this pr as it can grow uncontrolled. I'll open smaller ones. 🙌

**Comment by beltranrengifo** [comment]
ci:close
